packages/storage-plus/src/indexes.rs

```diff
@@ -39,31 +48,36 @@ where
     fn remove(&self, store: &mut dyn Storage, pk: &[u8], old_data: &T) -> StdResult<()>;
 }
 
-pub struct MultiIndex<'a, T> {
+pub struct MultiIndex<'a, K, T> {
     index: fn(&T) -> Vec<u8>,
-    idx_map: Map<'a, (&'a [u8], &'a [u8]), u32>,
-    // note, we collapse the pubkey - combining everything under the namespace - even if it is composite
+    idx_namespace: &'a [u8],
+    idx_map: Map<'a, (&'a [u8], &'a [u8]), T>,
+    // note, we collapse the pk - combining everything under the namespace - even if it is composite
     pk_map: Map<'a, &'a [u8], T>,
+    idx_type: PhantomData<K>,
 }
 
-impl<'a, T> MultiIndex<'a, T> {
+impl<'a, K, T> MultiIndex<'a, K, T> {
     // TODO: make this a const fn
     pub fn new(idx_fn: fn(&T) -> Vec<u8>, pk_namespace: &'a str, idx_namespace: &'a str) -> Self {
         MultiIndex {
             index: idx_fn,
             pk_map: Map::new(pk_namespace),
+            idx_namespace: idx_namespace.as_bytes(),
             idx_map: Map::new(idx_namespace),
+            idx_type: PhantomData,
         }
     }
 }
 
-impl<'a, T> Index<T> for MultiIndex<'a, T>
+impl<'a, K, T> Index<T> for MultiIndex<'a, K, T>
 where
     T: Serialize + DeserializeOwned + Clone,
+    K: PrimaryKey<'a>,
 {
     fn save(&self, store: &mut dyn Storage, pk: &[u8], data: &T) -> StdResult<()> {
         let idx = (self.index)(data);
-        self.idx_map.save(store, (&idx, &pk), &MARKER)
+        self.idx_map.save(store, (&idx, &pk), &data)
```

Comment:
We don't want to store all the data everywhere, as I need to update all indexes on any change, even when the index doesn't change. I put it in UniqueIndex as an optimization, but I think better to use MARKER there as well

Reply:
O know. See comment above. This is WIP, and I just did it to make things simpler at the beginning.